v1/ast/index.go

```diff
@@ -154,28 +155,60 @@ func (i *baseDocEqIndex) Lookup(resolver ValueResolver) (*IndexResult, error) {
 		return nil, err
 	}
 
-	result := NewIndexResult(i.kind)
+	result := IndexResultPool.Get()
+
+	result.Kind = i.kind
 	result.Default = i.defaultRule
 	result.OnlyGroundRefs = i.onlyGroundRefs
-	result.Rules = make([]*Rule, 0, len(tr.ordering))
+
+	if result.Rules == nil {
+		result.Rules = make([]*Rule, 0, len(tr.ordering))
+	} else {
+		result.Rules = result.Rules[:0]
+	}
+
+	clear(result.Else)
 
 	for _, pos := range tr.ordering {
-		sort.Slice(tr.unordered[pos], func(i, j int) bool {
-			return tr.unordered[pos][i].prio[1] < tr.unordered[pos][j].prio[1]
+		slices.SortFunc(tr.unordered[pos], func(a, b *ruleNode) int {
+			return a.prio[1] - b.prio[1]
 		})
 		nodes := tr.unordered[pos]
 		root := nodes[0].rule
 
 		result.Rules = append(result.Rules, root)
 		if len(nodes) > 1 {
+			if result.Else == nil {
+				result.Else = map[*Rule][]*Rule{}
+			}
+
 			result.Else[root] = make([]*Rule, len(nodes)-1)
 			for i := 1; i < len(nodes); i++ {
 				result.Else[root][i-1] = nodes[i].rule
 			}
 		}
 	}
 
-	result.EarlyExit = tr.values.Len() == 1 && tr.values.Slice()[0].IsGround()
+	if !tr.multiple {
+		// even when the indexer hasn't seen multiple values, the rule itself could be one
+		// where early exit shouldn't be applied.
```

Comment:
This comment is not describing new behaviour, but rather describes why we're doing this that we did before, right?

However, I think the `result.Rules[i].Head.DocKind() != CompleteDoc` check _is new_? It doesn't seem like that had been covered before. Do we need a new test case of some sort?

Reply:
This was something I struggled with for the longest time, and I wasn't sure this was even the right place to put this logic. I do believe it's covered by tests though, at least to the extent that all hell breaks loose if this is removed 😄